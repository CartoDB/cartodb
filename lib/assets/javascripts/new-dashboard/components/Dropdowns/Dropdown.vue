<template>
  <section v-click-outside="closeDropdown">
    <button class="dropdown__toggle" :class="{ 'dropdown__toggle--active': isOpen }" @click="toggleDropdown">
      <slot name="button"/>
    </button>

    <div class="dropdown" :class="{ 'is-open': isOpen }">
      <slot />
    </div>
  </section>
</template>

<script>
export default {
  name: 'Dropdown',
  data () {
    return {
      isOpen: false
    };
  },
  methods: {
    toggleDropdown () {
      this.isOpen = !this.isOpen;
    },
    closeDropdown () {
      this.isOpen = false;
    }
  }
};
</script>

<style lang="scss" scoped>
@import 'stylesheets/new-dashboard/variables';

.dropdown {
  visibility: hidden;
  position: absolute;
  z-index: 2;
  right: 0;
  width: 310px;
  margin-top: 8px;
  overflow: hidden;
  transition: all 0.25s linear;
  border: 1px solid $light-grey;
  border-radius: 2px;
  opacity: 0;
  pointer-events: none;

  &.is-open {
    visibility: visible;
    opacity: 1;
    pointer-events: auto;
  }
}

.dropdown__toggle {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 36px;
  padding: 0 9px;
  border-radius: 2px;

  &.dropdown__toggle--active {
    background-color: #F2F6F9;
  }
}
</style>
