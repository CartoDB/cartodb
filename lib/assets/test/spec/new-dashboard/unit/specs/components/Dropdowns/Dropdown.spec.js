import { mount } from '@vue/test-utils';
import Dropdown from 'new-dashboard/components/Dropdowns/Dropdown';

describe('Dropdown.vue', () => {
  it('should render correct contents', () => {
    const dropdown = mount(Dropdown, {
      slots: {
        default: 'Default text'
      }
    });
    expect(dropdown).toMatchSnapshot();
  });

  it('should open dropdown when clicking toggle', () => {
    const dropdown = mount(Dropdown, {
      slots: {
        default: 'Dropdown Content'
      }
    });

    dropdown.find('.dropdown__toggle').trigger('click');

    expect(dropdown).toMatchSnapshot();
  });

  describe('Methods', () => {
    describe('closeDropdown', () => {
      it('should close the dropdown', () => {
        const dropdown = mount(Dropdown, {
          data () { return { isOpen: true }; },
          slots: { default: 'Dropdown Content' }
        });

        let dropdownToggle = dropdown.find('.dropdown__toggle');
        let dropdownContent = dropdown.find('.dropdown');
        expect(dropdownToggle.classes()).toEqual(['dropdown__toggle', 'dropdown__toggle--active']);
        expect(dropdownContent.classes()).toEqual(['dropdown', 'is-open']);

        dropdown.vm.closeDropdown();

        expect(dropdownToggle.classes()).toEqual(['dropdown__toggle']);
        expect(dropdownContent.classes()).toEqual(['dropdown']);
      });
    });
  });
});
