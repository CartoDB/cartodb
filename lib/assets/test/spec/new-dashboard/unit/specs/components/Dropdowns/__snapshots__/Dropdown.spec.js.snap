// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown.vue should open dropdown when clicking toggle 1`] = `
<section><button class="dropdown__toggle dropdown__toggle--active"></button>
  <div class="dropdown is-open">Dropdown Content</div>
</section>
`;

exports[`Dropdown.vue should render correct contents 1`] = `
<section><button class="dropdown__toggle"></button>
  <div class="dropdown">Default text</div>
</section>
`;
